ksqldb-streams/src/main/java/io/confluent/ksql/execution/streams/PartitionByParamsFactory.java

```diff
@@ -202,6 +217,23 @@ private static LogicalSchema buildSchema(
     final String errorMsg = "Error computing new key from expression "
         + expressionMetadata.getExpression();
 
-    return row -> expressionMetadata.evaluate(row, null, logger, () -> errorMsg);
+    return new PartitionByExpressionEvaluator(
+        row -> expressionMetadata.evaluate(row, null, logger, () -> errorMsg),
+        isKeyExpression
+    );
+  }
+
+  private static class PartitionByExpressionEvaluator {
+
+    private final Function<GenericRow, Object> evaluator;
+    private final boolean acceptsKey;
```

Comment:
```suggestion
    private final boolean operatesOnKeyOnly;
```
A little hard for me to understand it inline without context on what this PR does:
```java
      final Object newKey = evaluator.evaluator.apply(
          evaluator.acceptsKey
          ? GenericRow.fromList(StructKeyUtil.asList(k))
          : v
      );
```

Reply:
I think this is addressed by your suggestion below? I went with `evaluateOnKeyOnly` though I don't feel strongly.